src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,230 @@
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+			context.RegisterOperationAction (operationAnalysisContext => {
+				var assignmentOperation = (IAssignmentOperation) operationAnalysisContext.Operation;
+				ProcessAssignmentOperation (operationAnalysisContext, assignmentOperation);
+			}, OperationKind.SimpleAssignment);
+
+			context.RegisterOperationAction (operationAnalysisContext => {
+				var invocationOperation = (IInvocationOperation) operationAnalysisContext.Operation;
+				ProcessInvocationOperation (operationAnalysisContext, invocationOperation);
+			}, OperationKind.Invocation);
+
+			context.RegisterOperationAction (operationAnalysisContext => {
+				var returnOperation = (IReturnOperation) operationAnalysisContext.Operation;
+				ProcessReturnOperation (operationAnalysisContext, returnOperation);
+			}, OperationKind.Return);
+		}
+
+		static void ProcessAssignmentOperation (OperationAnalysisContext context, IAssignmentOperation assignmentOperation)
+		{
+			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute))
+				return;
+
+			if (TryGetSymbolFromOperation (assignmentOperation.Target) is not ISymbol target ||
+				TryGetSymbolFromOperation (assignmentOperation.Value) is not ISymbol source)
+				return;
+
+			if (!target.TryGetDynamicallyAccessedMemberTypes (out var damtOnTarget))
+				return;
+
+			source.TryGetDynamicallyAccessedMemberTypes (out var damtOnSource);
+			if (Annotations.SourceHasRequiredAnnotations (damtOnSource, damtOnTarget, out var missingAnnotations))
+				return;
+
+			var diag = GetDiagnosticId (source.Kind, target.Kind);
+			var diagArgs = GetDiagnosticArguments (source.Kind == SymbolKind.NamedType ? context.ContainingSymbol : source, target, missingAnnotations);
+			context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), assignmentOperation.Syntax.GetLocation (), diagArgs));
+		}
+
+		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		{
+			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute))
+				return;
+
+			ProcessTypeArguments (context, invocationOperation);
+			ProcessArguments (context, invocationOperation);
+			if (!invocationOperation.TargetMethod.TryGetDynamicallyAccessedMemberTypes (out var damtOnCalledMethod))
+				return;
+
+			if (TryGetSymbolFromOperation (invocationOperation.Instance) is ISymbol instance &&
+				!instance.HasAttribute (RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute)) {
+				instance!.TryGetDynamicallyAccessedMemberTypes (out var damtOnCaller);
```

Comment:
This looks like `instance` will be the type of the "this" parameter - I think we want to be checking for annotations on the calling method, not the "this" type (which could be a different type entirely).

Reply:
Sorry, that was wrong - I think we want the annotation on the called method (which represents the annotation on its this parameter)